File: ingestion/validation.py
"""The Phase 0 validation gate.

Partitions chunks into loadable (complete Provenance Record) and flagged
(incomplete - never loaded), and checks structural integrity: no orphaned child
chunks, every parent link resolves, section gaps flagged (logged, not fatal).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Tuple

from ingestion.chunker import section_id
from ingestion.models import Chunk


@dataclass
class Flag:
    chunk_id: str
    reasons: List[str]


@dataclass
class ValidationReport:
    loadable: List[Chunk] = field(default_factory=list)
    flagged: List[Flag] = field(default_factory=list)
    orphaned_children: List[str] = field(default_factory=list)
    section_gaps: List[Tuple[str, str]] = field(default_factory=list)

    @property
    def structural_ok(self) -> bool:
        return not self.orphaned_children


def _section_gaps(chunks: List[Chunk]) -> List[Tuple[str, str]]:
    by_act: Dict[str, set] = {}
    for chunk in chunks:
        if chunk.section_number and chunk.section_number.isdigit():
            by_act.setdefault(chunk.act_id, set()).add(int(chunk.section_number))
    gaps: List[Tuple[str, str]] = []
    for act_id, numbers in by_act.items():
        for missing in range(min(numbers), max(numbers) + 1):
            if missing not in numbers:
                gaps.append((act_id, str(missing)))
    return gaps


def validate_chunks(chunks: List[Chunk]) -> ValidationReport:
    report = ValidationReport()

    valid_section_ids = {
        section_id(c.act_id, c.section_number)
        for c in chunks
        if c.section_number is not None
    }

    for chunk in chunks:
        if not chunk.is_loadable():
            report.flagged.append(
                Flag(chunk_id=chunk.chunk_id, reasons=chunk.provenance.missing_fields())
            )
            continue
        if chunk.is_child() and chunk.parent_section_id not in valid_section_ids:
            report.orphaned_children.append(chunk.chunk_id)
            continue
        report.loadable.append(chunk)

    report.section_gaps = _section_gaps(chunks)
    return report

File: web/scripts/lib/validation.ts
// The ingestion validation gate, ported from ingestion/validation.py.
// Partitions chunks into loadable (complete Provenance Record) and flagged
// (incomplete - never loaded), and checks structural integrity: no orphaned
// child chunks, every parent link resolves, section gaps flagged (logged,
// not fatal).
import {
  isLoadable,
  missingProvenanceFields,
  type Chunk,
} from "../../convex/lib/models";
import { sectionId } from "./chunker";

export type Flag = {
  chunkId: string;
  reasons: string[];
};

export type ValidationReport = {
  loadable: Chunk[];
  flagged: Flag[];
  orphanedChildren: string[];
  sectionGaps: Array<[string, string]>;
};

export function structuralOk(report: ValidationReport): boolean {
  return report.orphanedChildren.length === 0;
}

function sectionGapsOf(chunks: Chunk[]): Array<[string, string]> {
  const byAct = new Map<string, Set<number>>();
  for (const chunk of chunks) {
    if (chunk.sectionNumber && /^\d+$/.test(chunk.sectionNumber)) {
      if (!byAct.has(chunk.actId)) byAct.set(chunk.actId, new Set());
      byAct.get(chunk.actId)!.add(parseInt(chunk.sectionNumber, 10));
    }
  }
  const gaps: Array<[string, string]> = [];
  for (const [actId, numbers] of byAct) {
    const min = Math.min(...numbers);
    const max = Math.max(...numbers);
    for (let missing = min; missing <= max; missing++) {
      if (!numbers.has(missing)) gaps.push([actId, String(missing)]);
    }
  }
  return gaps;
}

export function validateChunks(chunks: Chunk[]): ValidationReport {
  const report: ValidationReport = {
    loadable: [],
    flagged: [],
    orphanedChildren: [],
    sectionGaps: [],
  };

  const validSectionIds = new Set(
    chunks
      .filter((c) => c.sectionNumber !== undefined)
      .map((c) => sectionId(c.actId, c.sectionNumber!)),
  );

  for (const chunk of chunks) {
    if (!isLoadable(chunk)) {
      report.flagged.push({
        chunkId: chunk.chunkId,
        reasons: missingProvenanceFields(chunk.provenance),
      });
      continue;
    }
    if (
      chunk.parentSectionId !== undefined &&
      !validSectionIds.has(chunk.parentSectionId)
    ) {
      report.orphanedChildren.push(chunk.chunkId);
      continue;
    }
    report.loadable.push(chunk);
  }

  report.sectionGaps = sectionGapsOf(chunks);
  return report;
}
